File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/AllBranchesShouldNotHaveSameImplementation.cs
using System;

namespace Tests.Diagnostics
{
    public class Program
    {
        public void IfElseCases(int b, int c)
        {
            if (b == 0)  // Noncompliant {{Remove this conditional structure or edit its code blocks so that they're not all the same.}}
//          ^^
            {
                DoSomething();
            }
            else if (b == 1)
            {
                DoSomething();
            }
            else
            {
                DoSomething();
            }

            if (b == 0) // Noncompliant
            {
                if (c == 1) // Noncompliant
//              ^^
                {
                    DoSomething();
                }
                else
                {
                    DoSomething();
                }
            }
            else
            {
                if (c == 1) // Noncompliant
                {
                    DoSomething();
                }
                else
                {
                    DoSomething();
                }
            }

            if (b == 0)
            {
                DoSomething();
            }
            else
            {
                DoSomethingElse();
            }

            if (b == 0) // Compliant - no else clause
            {
                DoSomething();
            }
            else if (b == 1)
            {
                DoSomething();
            }
        }

        public void SwitchCases(int i)
        {
            switch (i) // Noncompliant {{Remove this conditional structure or edit its code blocks so that they're not all the same.}}
//          ^^^^^^
            {
                case 1:
                    DoSomething();
                    break;
                case 2:
                    DoSomething();
                    break;
                case 3:
                    DoSomething();
                    break;
                default:
                    DoSomething();
                    break;
            }

            switch (i) // Noncompliant
            {
                case 1:
                    {
                        DoSomething();
                        break;
                    }
                case 2:
                    {
                        DoSomething();
                        break;
                    }
                case 3:
                    {
                        DoSomething();
                        break;
                    }
                default:
                    {
                        DoSomething();
                        break;
                    }
            }

            switch (i)
            {
                case 1:
                    DoSomething();
                    break;
                default:
                    DoSomethingElse();
                    break;
            }

            switch (i) // Compliant - no default section
            {
                case 1:
                    DoSomething();
                    break;
                case 2:
                    DoSomething();
                    break;
                case 3:
                    DoSomething();
                    break;
            }
        }

        public void TernaryCases(bool c, int a)
        {
            int b = a > 12 ? 4 : 4;  // Noncompliant {{This conditional operation returns the same value whether the condition is "true" or "false".}}
//                  ^^^^^^^^

            var x = 1 > 18 ? true : true; // Noncompliant
            var y = 1 > 18 ? true : false;
            y = 1 > 18 ? (true) : true; // Noncompliant
            TernaryCases(1 > 18 ? (true) : true, a); // Noncompliant
        }

        private void DoSomething()
        {
        }

        private void DoSomethingElse()
        {
        }

        public int SwitchExpressionNoncompliant(string type) =>
            type switch // Noncompliant {{Remove this conditional structure or edit its code blocks so that they're not all the same.}}
//               ^^^^^^
            {
                "a" => GetNumber(),
                "b" => GetNumber(),
                _ => GetNumber()
            };

        public int SwitchExpressionNested(string type)
        {
            return type switch
            {
                "a" => GetNumber(),
                _ => type switch // Noncompliant
                {
                        "b" => GetNumber(),
                        "c" => GetNumber(),
                        _ => GetNumber()
                    }
            };
        }

        public int GetNumber() => 42;

        public int SwitchExpressionCompliant(string type)
        {
            var x = type switch // Compliant
            {
                "a" => 42,
                "b" => type.Length,
                _ => GetNumber(),
            };
            string y = type switch { }; // Compliant
            var z = type switch { "a" => 42 }; // Compliant
            var withoutDefault = type switch // Compliant, does not have the discard default arm
            {
                "a" => GetNumber(),
                "b" => GetNumber(),
                "c" => GetNumber(),
                "d" => GetNumber(),
                "e" => GetNumber(),
            };
            return x;
        }
    }
}

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/AllBranchesShouldNotHaveSameImplementation.vb
Imports System

Namespace Tests.Diagnostics
    Public Class Program
        Public Sub IfElseCases(b As Integer, c As Integer)
            If b = 0 Then 'Noncompliant
'           ^^
                DoSomething()
            ElseIf b = 1 Then
                DoSomething()
            Else
                DoSomething()
            End If

            If b = 0 Then 'Noncompliant
                If c = 1 Then 'Noncompliant
                    DoSomething()
                Else
                    DoSomething()
                End If
            Else
                If c = 1 Then 'Noncompliant
                    DoSomething()
                Else
                    DoSomething()
                End If
            End If

            If b = 0 Then
                DoSomething()
            Else
                DoSomethingElse()
            End If

            If b = 0 Then 'Compliant, no else
                DoSomething()
            ElseIf b = 1 Then
                DoSomething()
            End If
        End Sub

        Public Sub SwitchCases(i As Integer)
            Select Case i ' Noncompliant {{Remove this conditional structure or edit its code blocks so that they're not all the same.}}
'           ^^^^^^
                Case 1
                    DoSomething()
                Case 2
                    DoSomething()
                Case 3
                    DoSomething()
                Case Else
                    DoSomething()
            End Select

            Select Case i ' Noncompliant
                Case 1
                    DoSomething()
                    Exit Select
                Case 2
                    DoSomething()
                    Exit Select
                Case 3
                    DoSomething()
                    Exit Select
                Case Else
                    DoSomething()
                    Exit Select
            End Select

            Select Case i
                Case 1
                    DoSomething()
                Case Else
                    DoSomethingElse()
            End Select

            Select Case i
                Case 1
                    DoSomething()
                Case 2
                    DoSomething()
                Case 3
                    DoSomething()
            End Select
        End Sub

        Public Sub TernaryCases(ByVal c As Boolean)
            Dim a As Integer = 1
            Dim b As Integer = If(a > 12, 4, 4) 'Noncompliant {{This conditional operation returns the same value whether the condition is "true" or "false".}}
'                              ^^
            Dim x = If(1 > 18, True, True) 'Noncompliant
            Dim y = If(1 > 18, True, False)
            y = If(1 > 18, (True), True) 'Noncompliant
            TernaryCases(If(1 > 18, (True), True)) 'Noncompliant
        End Sub

        Public Sub SingleLineIfCases(ByVal c As Boolean)
            Dim x As Integer
            If c Then x = 4 Else x = 4 'Noncompliant {{Remove this conditional structure or edit its code blocks so that they're not all the same.}}
'           ^^
            If c Then x = 0 Else x = 1
        End Sub

        Private Sub DoSomething()
        End Sub

        Private Sub DoSomethingElse()
        End Sub
    End Class
End Namespace
